### PR TITLE
STM32: Fix runtime device PM

### LINKED ISSUE #36917: Runtime device PM is broken on STM32
**Describe the bug**

@erwango reported that device PM related tests started to fail on STM32 targets after the introduction of device PM enum states cc2f0e9c08dcad59a4cc0117028a5fcf9d73f433. After some investigation, it turns out that the change to enums uncovered an existing issue on the STM32 pinmux driver.

Prior to the enum changes, we had:
https://github.com/zephyrproject-rtos/zephyr/blob/79a6c07536bc14583198f8e3555df6134d8822cf/include/pm/device.h#L27-L33

and we now have:
https://github.com/zephyrproject-rtos/zephyr/blob/a927ae39eda5ebc125437f3cee8fd741f0d161d5/include/pm/device.h#L27-L32

That is, the `PM_DEVICE_STATE_ACTIVE` was `1` and now is `0`. Making the enum start at `1` fixed the problem, something that sounded like an uninitialized variable.

The problem is the following:
1. UART(s) are initialized before GPIO drivers
2. UART makes use of the STM32 pinmux API to set up pinmux
3. The STM32 pinmux driver, in turn, makes use of the GPIO driver (PM only) without checking for it to be ready
https://github.com/zephyrproject-rtos/zephyr/blob/a927ae39eda5ebc125437f3cee8fd741f0d161d5/drivers/pinmux/pinmux_stm32.c#L129-L133
4. The GPIO driver pm_control function gets invoked, however, the initial pm_state is now `0` (static driver data is initialized to zeroes) which matches with the current value for `PM_DEVICE_STATE_ACTIVE`. Because the driver PM function only operates if there is a state change, the GPIO block is never enabled and so the serial port does not output anything, making the test to timeout. Note: If the test is debugged, one can verify that it passes, but nothing is _printed_.
https://github.com/zephyrproject-rtos/zephyr/blob/a927ae39eda5ebc125437f3cee8fd741f0d161d5/drivers/gpio/gpio_stm32.c#L614-L620

I'm not sure what is the best solution here, but:
1. Any device needs to be checked for readiness before operation. Making use of it without this check, even for PM, leads to undefined behavior.
2. Drivers should not care about the runtime PM status of other devices. In this case, either the pinmux driver needs to be PM-aware or it needs to use a GPIO API that will, in turn, enable/disable GPIO.

**To Reproduce**

Any test with `CONFIG_PM_DEVICE_RUNTIME=y` will fail because of a timeout (no serial output), e.g.

```
./scripts/twister -T tests/subsys/pm/power_mgmt --device-testing --device-serial /dev/ttyACM0 -p disco_l475_iot1
```

**Expected behavior**

Test passes as it does on other platforms (tested on native_posix and nrf52).

**Impact**

Runtime device PM is broken on STM32.

**Logs and console output**

N/A

**Environment (please complete the following information):**

 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: `main` or cc2f0e9c08dcad59a4cc0117028a5fcf9d73f433

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by erwango**
Runtime device PM was broken on STM32 following merge of https://github.com/zephyrproject-rtos/zephyr/commit/cc2f0e9c08dcad59a4cc0117028a5fcf9d73f433 that unveiled a flaw in current
STM32 PM implementation around gpio/pinmux.

Fix this issue by changing initialization step of gpio devices and reworking calls to gpio in pinmux
pseudo driver.

Fixes #36917

### CODE REVIEW

File: drivers/pinmux/pinmux_stm32.c [hunk 1]
```diff
@@ -46,15 +46,26 @@ const struct device * const gpio_ports[STM32_PORTS_MAX] = {
...
 		return -ENODEV;
 	}
 
-	return gpio_stm32_configure(port_device, STM32_PIN(pin), func, altf);
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	ret = pm_device_get(port_device);
+#endif
+	if (ret != 0) {
+		return ret;
+	}
```
Comment (ABOSTM) [inline drivers/pinmux/pinmux_stm32.c:60]: To be moved in #if above

File: drivers/pinmux/pinmux_stm32.c [hunk 2]
```diff
@@ -46,15 +46,26 @@ const struct device * const gpio_ports[STM32_PORTS_MAX] = {
...
+
+	gpio_stm32_configure(port_device, STM32_PIN(pin), func, altf);
```
Comment (ABOSTM) [inline drivers/pinmux/pinmux_stm32.c:62]: We lose the "return status" of `gpio_stm32_configure()` ?

**Review by ABOSTM — CHANGES_REQUESTED** [review]
typo in commit message of 1st commit:
di**z**sco_l475_iot1

File: drivers/pinmux/pinmux_stm32.c (hunk 1 above)
Comment (erwango) [reply drivers/pinmux/pinmux_stm32.c:60]: Indeed, thanks

File: drivers/pinmux/pinmux_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/pinmux/pinmux_stm32.c:62]: There is no error checking done in this function.
I wonder if it shouldn't be set to void.

Comment (erwango) [reply drivers/pinmux/pinmux_stm32.c:62]: Function moved to void

File: drivers/pinmux/pinmux_stm32.c [hunk 3]
```diff
@@ -54,10 +54,10 @@ static int stm32_pin_configure(uint32_t pin, uint32_t func, uint32_t altf)
 
 #ifdef CONFIG_PM_DEVICE_RUNTIME
 	ret = pm_device_get(port_device);
-#endif
 	if (ret != 0) {
 		return ret;
 	}
+#endif
```
Comment (ABOSTM) [inline drivers/pinmux/pinmux_stm32.c:60]: Fine, but fix should be part of 2nd commit

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: drivers/pinmux/pinmux_stm32.c (hunk 3 above)
Comment (erwango) [reply drivers/pinmux/pinmux_stm32.c:60]: oops! txs!

File: drivers/pinmux/pinmux_stm32.c [hunk 4]
```diff
@@ -46,15 +46,26 @@ const struct device * const gpio_ports[STM32_PORTS_MAX] = {
 static int stm32_pin_configure(uint32_t pin, uint32_t func, uint32_t altf)
 {
 	const struct device *port_device = gpio_ports[STM32_PORT(pin)];
```
Comment (gmarull) [inline drivers/pinmux/pinmux_stm32.c:48]: this function should actually do some bounds checking, ie

```c
const struct device *port_device;

if (STM32_PORT(pin) >= STM32_PORTS_MAX) {
        return -EINVAL;
}

port_device = gpio_ports[STM32_PORT(pin)];
```

File: drivers/gpio/gpio_stm32.c [hunk 5]
```diff
@@ -221,7 +221,7 @@ int gpio_stm32_configure(const struct device *dev, int pin, int conf, int altf)
 	z_stm32_hsem_unlock(CFG_HW_GPIO_SEMID);
 #endif  /* CONFIG_SOC_SERIES_STM32F1X */
 
-	return 0;
+	return;
```
Comment (gmarull) [inline drivers/gpio/gpio_stm32.c:224]: nit: no need to return

File: drivers/pinmux/pinmux_stm32.c (hunk 4 above)
Comment (erwango) [reply drivers/pinmux/pinmux_stm32.c]: Exact, done

**Review by gmarull — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM

**Pull request merged**